### ISSUE TITLE
`Oj::Parser` should support loading of multiple JSON elements 

### ISSUE THREAD

**Issue by gschlager** [ohler55/oj#930]
`Oj.load` can load multiple JSON elements from a stream. It would be great if `Oj::Parser.load` supported that as well.

> When used with a document that has multiple JSON elements the block, if any, will be yielded to. If no block then the last element read will be returned.

**Comment by ohler55** [comment]
Oj::Parser is mostly prepared to add that feature. I'll add it.

**Comment by ohler55** [comment]
Try the parser-multi-load branch. I added a test to the test_parser_usual.rb file. The branch still needs some testing but so far it works.

**Comment by gschlager** [comment]
Thanks! We might have been talking about a different feature. Here's some example code:

```ruby
reader, writer = IO.pipe
writer.write('{"a":1}')
writer.write('{"b":{"x":2}}')
writer.write('{"c":3}')
writer.close

Oj.load(reader) { |data| puts data }
reader.close
```

This outputs the following:
```
{"a"=>1}
{"b"=>{"x"=>2}}
{"c"=>3}
```

Ideally, I'd like to do the same with `Oj::Parser`, but the following doesn't work right now.

```ruby
p = Oj::Parser.new(:usual)
p.load(reader) { |data| puts data }
```

**Comment by ohler55** [comment]
Updated the branch. Off by one. There is still an issue to be resolved with reusing the parser a second time. It should get solved with more testing.

We are in agreement on the feature. I works (should) with both parse and load.

**Comment by gschlager** [comment]
I might be doing something wrong because it doesn't work for me.

Here's a basic example with `Oj.load` that works:

```ruby
#!/usr/bin/env ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "oj", github: "ohler55/oj", branch: "parser-multi-load"
end

reader, writer = IO.pipe

thread =
  Thread.new do
    5.times do |id|
      Oj.to_stream(writer, { "id" => id })
      sleep(1)
    end

    writer.close
  end

Oj.load(reader) { |data| puts "#{Time.now} -- ID: #{data["id"]}" }

reader.close
thread.join
```

Output:
```
2024-08-01 15:38:00 +0000 -- ID: 0
2024-08-01 15:38:01 +0000 -- ID: 1
2024-08-01 15:38:02 +0000 -- ID: 2
2024-08-01 15:38:03 +0000 -- ID: 3
2024-08-01 15:38:04 +0000 -- ID: 4
```

---

And here's the same with `Oj::Parser`. It doesn't work for me.

```ruby
#!/usr/bin/env ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "oj", github: "ohler55/oj", branch: "parser-multi-load"
end

reader, writer = IO.pipe

thread =
  Thread.new do
    5.times do |id|
      Oj.to_stream(writer, { "id" => id })
      sleep(1)
    end

    writer.close
  end

p = Oj::Parser.new(:usual)
p.load(reader) { |data| puts "#{Time.now} -- ID: #{data["id"]}" }

reader.close
thread.join
```

Output:
```
2024-08-01 15:38:55 +0000 -- ID: 0
```

**Comment by ohler55** [comment]
Did you look at the `test_multi` test in the test/test_parser_usual.rb file?

I'll try the delayed writes and make sure they work tonight.

**Comment by ohler55** [comment]
Made the fix and added a test last night. Please give it a try.

**Comment by ohler55** [comment]
Released v3.16.5 with the support for multiple loading.
